### ISSUE TITLE
An idea for a major refactoring of the audio code

### ISSUE THREAD

**Issue by kirbyfan64** [Secretchronicles/TSC#590]
This should fix all currently open bugs related to audio (and probably some that haven't been discovered yet!). This would also be entirely API-compatible.

TL;DR: my idea is to rewrite `cAudio` to use threads. There will be the main thread as usual, but there will also be a newly-spawned "manager" thread. When someone does something like asking the music to fade, the main thread will simply place a "message" in a mutex-guarded queue. The manager thread will be constantly waiting for new messages and executing the corresponding actions.

That doesn't sound like much of an improvement, but it really is. The code is currently pretty much a demented state machine, and I've seen a few bugs that were caused by implicit assumptions regarding that state. This whole thing would basically remove the need for crud like all the "Is_Music_XXX" checks in the code

I know this isn't that convincing, but I just want to get the idea out there.

**Comment by datahead8888** [comment]
If we do the complete rewrite of the game, you might want to consider how much rewriting you want to do now or whether it makes sense to wait and do it in the newer API for the complete rewrite.  If there is no other feasible way to fix the bugs, it will of course have to be done in version 2.1 regardless.

If we instead do an incremental rewrite, fire ahead, as long as you are certain it should make it more stable and not as prone to bugs for the next 2.1 release.  Either way we go with the rewrites, these systems will have to be rewritten one way or the other; it's just a question of when and how.

The discussion on complete or incremental rewriting is here: http://forum.secretchronicles.de/forums/2/threads/2969 -- could you post to the thread when you get a chance?

**Comment by Quintus** [comment]
Suggestion. @kirbyfan64 before you start a rewrite here, give us your opinion on entire TSC rewrite at the link datahead gave. We don't want to stop you, but we want to have an overview of the opinions in the TSC project (please note that if you reply by mail that we have a `.org` domain now and emails to `.de` get rejected).

Vale,
Quintus

**Comment by datahead8888** [comment]
> We don't want to stop you

To Quintus's point, some sort of progress is always better than no progress - we will at least learn something if you work with this now, if not make direct progress towards the rewrite.  We do, though, want your input on the rewrite method for the game as a whole.

**Comment by datahead8888** [comment]
@kirbyfan64 - An incremental rewrite of TSC is a lot more likely than it was before, so if you're still up for doing this audio system revamp -- fire ahead!  Fixing the existing audio bugs will help us stabilize version 2.1, and having more supportable audio code will also be a big win.

**Comment by datahead8888** [comment]
@kirbyfan64 - do you think you still might want to do the revamp of the audio system to use threads?  You had said it might be a good way to knock out the remaining audio bugs.  Bugs are the primary impediment to finishing release 2.1, so any bug fix is a big win.

**Comment by kirbyfan64** [comment]
Yeah, I just lost a bit of enthusiasm after losing the code three times in a row. :/

**Comment by datahead8888** [comment]
> Yeah, I just lost a bit of enthusiasm after losing the code three times in a row. :/

Losing the code three times in a row?  That's a lot of losses - you might look into data recovery options depending on how you lost it.

Commit any code you write in a branch on a day to day basis, and upload it to GitHub in this repository to back it up.  Use as many commits as you need to (without even squashing them).  The activity in GitHub will also be good for the project.

**Comment by datahead8888** [comment]
@kirbyfan64, do you know if it might be more feasible to fix the audio bugs without doing a major refactoring of the code?  If we at least get them fixed, it knocks out one of the big roadblocks for releasing version 2.1.

If you remember what you had written on the refactoring and can do it quickly again, this also could be a good solution, but getting the bugs fixed is ultimately the critical item.
